discovery_engine_core/providers/src/bing.rs

```diff
@@ -0,0 +1,139 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! Client to retrieve trending topics.
+
+use serde::{Deserialize, Deserializer, Serialize};
+use url::Url;
+
+use crate::{seal::Seal, Client, Error, Market, Query};
+
+const URL_SUFFIX: &str = "_tt";
+
+/// Parameters for fetching trending news topics.
+pub struct TrendingQuery<'a> {
+    /// Market to fetch results from.
+    pub market: &'a Market,
+}
+
+impl Query for TrendingQuery<'_> {
+    fn setup_url(&self, url: &mut Url) -> Result<(), Error> {
+        url.path_segments_mut()
+            .map_err(|_| Error::InvalidUrlBase(None))?
+            .push(URL_SUFFIX);
+
+        let query = &mut url.query_pairs_mut();
+
+        let country = &self.market.country_code;
+        let lang = &self.market.lang_code;
+        query.append_pair("mkt", &format!("{}-{}", lang, country));
+
+        Ok(())
+    }
+}
+
+impl Seal for TrendingQuery<'_> {}
+
+impl Client {
+    /// Run query for fetching trending topics from Bing.
+    pub async fn query_trending(&self, query: &impl Query) -> Result<Vec<TrendingTopic>, Error> {
+        self.query_bing(query).await.map(|trending| trending.value)
+    }
+
+    /// Run a query against Bing.
+    pub async fn query_bing(&self, query: &impl Query) -> Result<Response, Error> {
+        let mut url = Url::parse(&self.url).map_err(|e| Error::InvalidUrlBase(Some(e)))?;
+        query.setup_url(&mut url)?;
+
+        let response = self
+            .client
+            .get(url)
+            .timeout(self.timeout)
+            .bearer_auth(&self.token)
+            .send()
+            .await
+            .map_err(Error::RequestExecution)?
+            .error_for_status()
+            .map_err(Error::StatusCode)?;
+
+        let raw_response = response.bytes().await.map_err(Error::Fetching)?;
+        let deserializer = &mut serde_json::Deserializer::from_slice(&raw_response);
+        serde_path_to_error::deserialize(deserializer)
+            .map_err(|error| Error::DecodingAtPath(error.path().to_string(), error))
+    }
+}
+
+/// Query response from Bing API.
+#[derive(Deserialize, Debug)]
+pub struct Response {
+    /// Main response content.
+    #[serde(default)]
+    value: Vec<TrendingTopic>,
+}
+
+/// Trending topic.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TrendingTopic {
+    /// Title of the trending topic.
+    #[serde(default, deserialize_with = "deserialize_null_default")]
+    pub(crate) name: String,
+
+    /// Search query that returns this topic.
+    #[serde(default)]
+    pub(crate) query: SearchQuery,
+
+    /// Link to a related image.
+    #[serde(default)]
+    pub(crate) image: Image,
```

Comment:
❓ are the `SearchQuery` and `Image` newtypes actually required for the serde? it looks like we could just make them `String`s  and put the respective serde attributes on them which are currently within the newtypes.

Reply:
they seem to be necessary (did try the suggestion, adjusting the tests, but they break). I suppose it is because, for example under `image` in the json there are fields `url` and `provider` and the deserializer needs to be told which one we want. perhaps someone who knows more serde knows a more economical way to do it...